api/preProcessor-template.ts

```diff
@@ -0,0 +1,19 @@
+import thing from 'module';
+
+export let sourceExtensionTypes: string[] = [".ts", ".tsx"];
+
+export let outputExtensionTypes: string[] = [".mjs", ".cjs"];
+
+
+// Example of defining "options" for a TypeScript loader
+export interface TypeScriptOptions {
+  tsconfigFile?: string
+}
+
+export function process(source: string | Buffer, options: TypeScriptOptions = {}): string| Buffer {
```

Comment:
One possible issue with the current API is that it forces the preprocessor to be entirely stateless. Which means that it becomes tricky to reuse the same TS compiler host for example. To do that, the processor would be forced to somehow derive a key from the `options` and keep a map of configured compiler hosts.

One alternative would be to instead expose a factory, roughly:

```ts
export function getPreprocessor(options: TypeScriptOptions) {
  // create compiler host etc. with options.

  return {
    process(source: string) {
      // compile the source, reusing state from previous compilations.
    },
  };
}
```

An alternate style is to expose something taking a builder/registry which could also incorporate the other info:

```ts
export function getPreprocessor(builder, options: TypeScriptOptions) {
  // create compiler host etc. with options.

  builder
    .fromExtensions('.ts', '.tsx')
    .toExtensions('.mjs', '.cjs')
    .transform(source => {
      // compile the source, reusing state from previous compilations.
      return { extension: '.mjs', source: /* ... */ };
    });
}
```

Reply:
Another good call!  I really like the idea of a "factory" - being able to reuse the state created once via this factory makes intuitive sense to me.  I'll look more into the builder/registry method, however, to be able to compare before a further decision.

Reply:
I've done some testing around both the "factory" and the "builder" design patterns.  I personally like the "factory" idea more to allow for a stateful `preProcessor` - I believe this same method should be used for `postProcessors`.  A future consideration is the use of threads/workers for these processors, but I will explore this down the line.